Implement trusted types enforcement on eval

lukewarlow opened the pull request:
#### 7a4725b0e7299c17437ca3c94836d430bd741aac
<pre>
Implement trusted types enforcement on eval
<a href="https://bugs.webkit.org/show_bug.cgi?id=273185">https://bugs.webkit.org/show_bug.cgi?id=273185</a>

Reviewed by Darin Adler and Justin Michaud.

This patch introduces a new canCompileStrings function to the global object method table.
This function is used in eval for enforcement of trusted types.
It also updates the associated tests to match the latest spec.

A follow up patch will update the Function constructor to use this.

See <a href="https://tc39.es/proposal-dynamic-code-brand-checks">https://tc39.es/proposal-dynamic-code-brand-checks</a>

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-importScripts-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/csp-block-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-no-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/WorkerGlobalScope-eval.https.js:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-expected.txt: Added.
* Source/JavaScriptCore/API/JSAPIGlobalObject.cpp:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::baseGlobalObjectMethodTable):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::canCompileStrings):
(JSC::JSGlobalObject::requiresTrustedTypes const):
(JSC::JSGlobalObject::setRequiresTrustedTypes):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::canCompileStrings):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::globalObjectMethodTable):
(WebCore::JSDOMWindowBase::canCompileStrings):
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
(WebCore::JSShadowRealmGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSShadowRealmGlobalScopeBase::canCompileStrings):
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
(WebCore::JSWorkerGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkerGlobalScopeBase::canCompileStrings):
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.cpp:
(WebCore::JSWorkletGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkletGlobalScopeBase::canCompileStrings):
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::setRequiresTrustedTypes):
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setRequiresTrustedTypes):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::canCompile):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::didCreateWindowProxy const):
(WebCore::ContentSecurityPolicy::applyPolicyToScriptExecutionContext):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.cpp:
(WebCore::WorkerOrWorkletGlobalScope::setRequiresTrustedTypes):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::setRequiresTrustedTypes):
* Source/WebCore/workers/WorkerOrWorkletScriptController.h:

Canonical link: <a href="https://commits.webkit.org/279473@main">https://commits.webkit.org/279473@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a9411137fcd14ec56631a6701232112f27e0843

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56789 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4235 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55815 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40329 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4026 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43377 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2776 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55607 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31064 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46260 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3563 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2391 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46869 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3735 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58384 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53024 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28667 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3744 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50783 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46429 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50123 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11678 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30802 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65329 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29645 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12451 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -476,20 +476,31 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
+    String s = codeString->value(globalObject);
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
+
+    if (!globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::IndirectEval, s, x)) {
```
TODO: Add a boolean to the global object to indicate TT enforcement (report only or fully enforced), and only then do the call through to webcore.

lukewarlow replied:
Do the same to the other file

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate.html:
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html><!-- webkit-test-runner [ jscOptions=--useTrustedTypes=true ] -->
```
Exported at https://github.com/web-platform-tests/wpt/pull/46484

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate-expected.txt:
```diff
@@ -0,0 +1,5 @@
+
+PASS eval of string where default policy mutates value throws.
+PASS indirect eval of string where default policy mutates value throws.
+FAIL Function constructor with string where default policy mutates value throws. assert_throws_js: function "_ => new Function('return 1+1')" did not throw
```
Why is this failing?

lukewarlow replied:
That isn't addressed by this PR a followup will address that functionality.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
```diff
@@ -1,17 +1,15 @@
-FAIL new Function with default policy and permissive CSP still obeys default policy. assert_equals: expected "Hello a cat untrusted string" but got "Hello transformed untrusted string"
+PASS eval with plain string with Trusted Types and permissive CSP works with default policy.
+PASS indirect eval with plain string with Trusted Types and permissive CSP works with default policy.
+FAIL new Function with plain string default policy and permissive CSP works with default policy. assert_equals: expected (number) 2 but got (undefined) undefined
```
Why is this failing?

lukewarlow replied:
Good catch that's a mistake in the test. Will fix it and add it to the export.

darinadler commented on Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
```diff
@@ -45,6 +45,8 @@ enum class JSPromiseRejectionOperation : unsigned {
     Handle, // When a handler is added to a rejected promise for the first time.
 };
 
+enum class CompilationType;
```
This forward declaration should be sharing a paragraph with the other forward declaration, of `ScriptExecutionStatus`, not after the declaration of `JSPromiseRejectionOperation`.

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -551,6 +557,7 @@ class JSGlobalObject : public JSSegmentedVariableObject {
 
     bool m_evalEnabled { true };
     bool m_webAssemblyEnabled { true };
+    bool m_requireTrustedTypes { true };
```
Naming style here would be `m_requiresTrustedTypes`.

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -1035,6 +1044,10 @@ class JSGlobalObject : public JSSegmentedVariableObject {
         m_webAssemblyEnabled = enabled;
         m_webAssemblyDisabledErrorMessage = errorMessage;
     }
+    void setRequireTrustedTypes(bool required)
```
Naming style here would be `setRequiresTrustedTypes`.

darinadler commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -309,6 +312,25 @@ String JSDOMWindowBase::codeForEval(JSGlobalObject* globalObject, JSValue value)
+    if (result.hasException()) {
+        propagateException(*globalObject, throwScope, result.releaseException());
+        RETURN_IF_EXCEPTION(throwScope, false);
+        return false;
```
Seems like this is unreachable code.

darinadler commented on Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
```diff
@@ -156,6 +159,25 @@ String JSWorkerGlobalScopeBase::codeForEval(JSC::JSGlobalObject* globalObject, J
+    VM& vm = globalObject->vm();
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    JSWorkerGlobalScopeBase& thisObject = static_cast<JSWorkerGlobalScopeBase&>(*globalObject);
```
I suggest `auto&` here to avoid writing the type out twice.

darinadler commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -309,6 +312,25 @@ String JSDOMWindowBase::codeForEval(JSGlobalObject* globalObject, JSValue value)
+    VM& vm = globalObject->vm();
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    JSDOMWindowBase& thisObject = static_cast<JSDOMWindowBase&>(*globalObject);
```
I suggest `auto&` here to avoid writing the type out twice.

darinadler commented on Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
```diff
@@ -156,6 +159,25 @@ String JSWorkerGlobalScopeBase::codeForEval(JSC::JSGlobalObject* globalObject, J
     return nullString();
 }
 
+bool JSWorkerGlobalScopeBase::canCompileStrings(JSC::JSGlobalObject* globalObject, JSC::CompilationType compilationType, String codeString, JSC::JSValue bodyArgument)
```
Really annoying that this code is basically exactly the same as `JSDOMWindowBase::canCompileStrings`. Can we share more code?

lukewarlow replied:
I'll have a look to see if I can move it to a shared function somewhere

darinadler commented on Source/WebCore/bindings/js/ScriptController.cpp:
```diff
@@ -453,6 +453,14 @@ void ScriptController::setWebAssemblyEnabled(bool value, const String& errorMess
 
+void ScriptController::setRequireTrustedTypes(bool required)
+{
+    auto* jsWindowProxy = windowProxy().existingJSWindowProxy(mainThreadNormalWorld());
```
I suggest we call this local variable just `proxy`.

darinadler commented on Source/WebCore/bindings/js/ScriptController.h:
```diff
@@ -130,6 +130,7 @@ class ScriptController final : public CanMakeWeakPtr<ScriptController>, public C
 
     void setEvalEnabled(bool, const String& errorMessage = String());
     void setWebAssemblyEnabled(bool, const String& errorMessage = String());
+    void setRequireTrustedTypes(bool);
```
`setRequiresTrustedTypes`

darinadler commented on Source/WebCore/dom/TrustedType.h:
```diff
@@ -30,6 +30,14 @@
+
+enum class CompilationType;
+
+class JSValue;
```
We normally sort these alphabetically, so `class` comes before `enum class`.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -300,6 +302,10 @@ void ContentSecurityPolicy::applyPolicyToScriptExecutionContext()
         if (policy->hasBlockAllMixedContentDirective() && !policy->isReportOnly())
             enableStrictMixedContentMode = true;
+
+        // Intentionally doesn't check for report only, this check is used for performance purposes.
```
I don’t think you mean "this check is used for performance purposes". Maybe you mean something more like "this check can be skipped for higher performance"?

lukewarlow replied:
I'll rewrite that comment to clarify it. I meant this doesn't check for report only because the Boolean this creates is used later for performance purposes not for CSP enforcement.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -292,6 +293,7 @@ void ContentSecurityPolicy::applyPolicyToScriptExecutionContext()
     updateSourceSelf(*m_scriptExecutionContext->securityOrigin());
 
     bool enableStrictMixedContentMode = false;
+    bool requireTrustedTypesForScript = false;
```
`requires` instead of `require`

darinadler commented on Source/WebCore/workers/WorkerOrWorkletScriptController.h:
```diff
@@ -95,6 +95,7 @@ class WorkerOrWorkletScriptController final : public CanMakeCheckedPtr<WorkerOrW
 
     void disableEval(const String& errorMessage);
     void disableWebAssembly(const String& errorMessage);
+    void setRequireTrustedTypes(bool required);
```
`setRequiresTrustedTypes`

justinmichaud reviewed (approved):
The JSC side looks good to me too

lukewarlow committed:
Implement trusted types enforcement on eval
https://bugs.webkit.org/show_bug.cgi?id=273185

Reviewed by Darin Adler and Justin Michaud.

This patch introduces a new canCompileStrings function to the global object method table.
This function is used in eval for enforcement of trusted types.
It also updates the associated tests to match the latest spec.

A follow up patch will update the Function constructor to use this.

See https://tc39.es/proposal-dynamic-code-brand-checks

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-importScripts-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/csp-block-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-no-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/WorkerGlobalScope-eval.https.js:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-expected.txt: Added.
* Source/JavaScriptCore/API/JSAPIGlobalObject.cpp:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::baseGlobalObjectMethodTable):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::canCompileStrings):
(JSC::JSGlobalObject::requiresTrustedTypes const):
(JSC::JSGlobalObject::setRequiresTrustedTypes):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::canCompileStrings):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::globalObjectMethodTable):
(WebCore::JSDOMWindowBase::canCompileStrings):
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
(WebCore::JSShadowRealmGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSShadowRealmGlobalScopeBase::canCompileStrings):
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
(WebCore::JSWorkerGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkerGlobalScopeBase::canCompileStrings):
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.cpp:
(WebCore::JSWorkletGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkletGlobalScopeBase::canCompileStrings):
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::setRequiresTrustedTypes):
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setRequiresTrustedTypes):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::canCompile):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::didCreateWindowProxy const):
(WebCore::ContentSecurityPolicy::applyPolicyToScriptExecutionContext):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.cpp:
(WebCore::WorkerOrWorkletGlobalScope::setRequiresTrustedTypes):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::setRequiresTrustedTypes):
* Source/WebCore/workers/WorkerOrWorkletScriptController.h:

Canonical link: https://commits.webkit.org/279473@main
